source/http/HttpConnectionManager.cpp

```diff
@@ -0,0 +1,349 @@
+                                m_pendingConnectionRequests.pop_front();
+                                m_lastError = aws_last_error();
+                                callbackError = m_lastError;
+                            }
+                        }
+                        connection = nullptr;
+                        /* Case 1.1 */
+                    }
+
+                    /* Case 2.1*/
+                    if (poolOrVend && !m_pendingConnectionRequests.empty())
+                    {
+                        userCallback = m_pendingConnectionRequests.front();
+                        m_pendingConnectionRequests.pop_front();
+                        ++m_outstandingVendedConnections;
+                        callbackError = AWS_ERROR_SUCCESS;
+                        vendedConnection = std::move(connection);
+                    }
+                    else if (poolOrVend)
+                    {
+                        /* Case 2.2 */
+                        m_connections.push_back(std::move(connection));
+                    }
+                }
+
+                if (userCallback)
+                {
+                    userCallback(vendedConnection, callbackError);
+                }
+            }
+
+            void HttpClientConnectionManager::ReleaseConnection(
+                std::shared_ptr<HttpClientConnection> connection) noexcept
+            {
+                poolOrVendConnection(std::move(connection), true);
+            }
+
+            void HttpClientConnectionManager::onConnectionSetup(
+                const std::shared_ptr<HttpClientConnection> &connection,
+                int errorCode) noexcept
+            {
+                if (!errorCode && connection)
+                {
+                    poolOrVendConnection(connection, false);
+                    return;
+                }
+
+                /* this only gets hit if the connection setup failed. */
+                Vector<std::pair<OnClientConnectionAvailable, int>> userCallbackNewConnectionFailures;
+                {
+                    std::lock_guard<std::mutex> connectionsLock(m_connectionsLock);
+                    --m_pendingConnections;
+                    if (!m_pendingConnectionRequests.empty())
+                    {
+                        userCallbackNewConnectionFailures.push_back(std::pair<OnClientConnectionAvailable, int>(
+                            m_pendingConnectionRequests.front(), errorCode));
+                        m_pendingConnectionRequests.pop_front();
+                    }
+
+                    /* if we had more pending connection requests than the pool size, we need to replace this
+                       connection. Just loop until we've matched either the connectionRequest to the max amount of
+                       connections in flight, or when run out of pending connection requests. */
+                    while (m_pendingConnectionRequests.size() > m_pendingConnections &&
```

Comment:
This case is  **very** similar to the "case 1.2" in `poolOrVendConnection()`, but we handle it differently here in `onConnectionSetup()`

In `poolOrVendConnection()` we try **once** to create a new connection to replace the broken one. Here in `onConnectionSetup()` we try **repeatedly** to create a new connection to replace the broken one.

I have a hunch that trying just once is the better approach, if for no other reason than it's simpler. But maybe there's some complication I'm not thinking of?